jupyterhub_moss/spawner.py

```diff
@@ -361,57 +383,20 @@ def options_from_form(self, formdata: Dict[str, List[str]]) -> Dict[str, str]:
             self.req_prologue, options["environment_path"], partition_environments
         )
 
-        return options
-
-    def __check_user_options(self, partition_info):
-        """Check if requested resources are valid for the given partition info.
-
-        See :meth:`__validate_options` for the other user options checks.
-
-        Raises AssertionError if request does not match available resources.
-        """
-        if "runtime" in self.user_options:
-            runtime = parse_timelimit(self.user_options["runtime"])
-            assert (
-                runtime.total_seconds() <= partition_info["max_runtime"]
-            ), "Requested runtime exceeds partition time limit"
-
-        if (
-            "nprocs" in self.user_options
-            and self.user_options["nprocs"] > partition_info["max_nprocs"]
-        ):
-            raise AssertionError("Error in number of CPUs")
-
-        if (
-            "ngpus" in self.user_options
-            and self.user_options["ngpus"] > partition_info["max_ngpus"]
-        ):
-            raise AssertionError("Error in number of GPUs")
+    async def options_from_form(self, formdata: Dict[str, List[str]]) -> Dict[str, str]:
+        """Parse the form and add options to the SLURM job script"""
+        options = self.__convert_formdata(formdata)
 
-    def __update_spawn_options(self, partition_info):
-        """Update user_options and other attributes controlling the spawn"""
-        # Specific handling of exclusive flag
-        # When mem=0 or all CPU are requested, set the exclusive flag
-        if (
-            self.user_options.get("nprocs") == partition_info["max_nprocs"]
-            or self.user_options.get("mem") == "0"
-        ):
-            self.user_options["exclusive"] = True
+        assert "partition" in options, "Partition information is missing"
+        assert options["partition"] in self.partitions, "Partition is not supported"
+        partitions_info = await self._get_partitions_info()
+        partition_info = partitions_info[options["partition"]]
 
-        # Specific handling of landing URL (e.g., to start jupyterlab)
-        self.default_url = self.user_options.get("default_url", "")
-        self.log.info(f"Used default URL: {self.default_url}")
+        self.__validate_options(options, partition_info)
 
-        if "root_dir" in self.user_options:
-            self.notebook_dir = self.user_options["root_dir"]
+        self.__update_options(options, partition_info)
```

Comment:
Updating options here to request `partition_info` only once since this calls an external tool and might be costly.